File: libs/caramba-core/csm/cbor.go
package csm

import "unicode/utf8"

// Строгий профиль декодирования CBOR, 03-WIRE.md раздел 3.
//
// Каноничность в CSM/1 это локальный предикат над входными байтами, а не
// сравнение с результатом перекодирования. Декодер написан здесь, а не взят
// из библиотеки, потому что профиль это подмножество: библиотека общего
// назначения примет то, что профиль обязан отвергнуть (неопределённые длины,
// теги, float, null, отрицательные числа, неминимальные головы, дубли ключей).
//
// Все ограничения проверяются ПО ХОДУ разбора, до выделения памяти, чтобы
// враждебный документ не мог заставить нас выделить буфер по своему слову.

// Ограничения раздела 3.2.
const (
	MaxDepth      = 6
	MaxMapPairs   = 64
	MaxArrayItems = 512
	MaxTstrBytes  = 256
	MaxBstrBytes  = 3072
	MaxUint       = uint64(1)<<53 - 1
)

// Диапазоны ключей раздела 3.3.
const (
	CriticalKeyMin    = 1
	CriticalKeyMax    = 63
	NonCriticalKeyMin = 64
	NonCriticalKeyMax = 1023
)

// Kind это тип элемента CBOR, допустимый профилем.
type Kind uint8

const (
	KindUint Kind = iota
	KindBstr
	KindTstr
	KindArray
	KindMap
	KindBool
)

func (k Kind) String() string {
	switch k {
	case KindUint:
		return "uint"
	case KindBstr:
		return "bstr"
	case KindTstr:
		return "tstr"
	case KindArray:
		return "array"
	case KindMap:
		return "map"
	case KindBool:
		return "bool"
	}
	return "?"
}

// Pair это одна пара карты. Ключ всегда беззнаковое целое (правило C9).
type Pair struct {
	Key uint64
	Val Value
}

// Value это разобранный элемент CBOR. Bstr и Tstr указывают внутрь входного
// буфера и не копируются: проверка подписи идёт по принятым байтам, и лишняя
// копия только даёт шанс разойтись с ними.
type Value struct {
	Kind  Kind
	U     uint64
	B     []byte
	S     string
	Bool  bool
	Array []Value
	Map   []Pair // строго по возрастанию ключа, правило C10
}

// Get возвращает значение по ключу карты. Пары отсортированы по возрастанию,
// поэтому поиск линейный с ранним выходом.
func (v *Value) Get(key uint64) (*Value, bool) {
	if v.Kind != KindMap {
		return nil, false
	}
	for i := range v.Map {
		if v.Map[i].Key == key {
			return &v.Map[i].Val, true
		}
		if v.Map[i].Key > key {
			break
		}
	}
	return nil, false
}

// Has сообщает, присутствует ли ключ.
func (v *Value) Has(key uint64) bool {
	_, ok := v.Get(key)
	return ok
}

type cborReader struct {
	b []byte
	i int
}

func cborErr(format string, args ...any) *Error {
	return errf(EParseCBOR, "P9", format, args...)
}

// decodeCBORPayload разбирает полезную нагрузку кадра: ровно один элемент
// верхнего уровня, обязательно карта (C1), потребляющий ровно len(p) байт (C2).
func decodeCBORPayload(p []byte) (Value, error) {
	r := &cborReader{b: p}
	v, err := r.item(1)
	if err != nil {
		return Value{}, err
	}
	if v.Kind != KindMap {
		return Value{}, cborErr("C1: top-level item is %s, must be a map", v.Kind)
	}
	if r.i != len(p) {
		return Value{}, cborErr("C2: %d trailing byte(s) inside payload_len", len(p)-r.i)
	}
	return v, nil
}

// remaining возвращает число непрочитанных байт.
func (r *cborReader) remaining() int { return len(r.b) - r.i }

// head читает начальный байт и его аргумент, проверяя правила C3 (только
// определённые длины) и C4 (кратчайшая форма головы).
func (r *cborReader) head() (major byte, ai byte, arg uint64, err error) {
	if r.remaining() < 1 {
		return 0, 0, 0, cborErr("truncated: expected an item head")
	}
	ib := r.b[r.i]
	r.i++
	major = ib >> 5
	ai = ib & 0x1f

	switch {
	case ai < 24:
		arg = uint64(ai)
	case ai == 24:
		if r.remaining() < 1 {
			return 0, 0, 0, cborErr("truncated one-byte argument")
		}
		arg = uint64(r.b[r.i])
		r.i++
		if arg < 24 {
			return 0, 0, 0, cborErr("C4: non-minimal head, %d encoded with a one-byte argument", arg)
		}
	case ai == 25:
		if r.remaining() < 2 {
			return 0, 0, 0, cborErr("truncated two-byte argument")
		}
		arg = uint64(r.b[r.i])<<8 | uint64(r.b[r.i+1])
		r.i += 2
		if arg < 256 {
			return 0, 0, 0, cborErr("C4: non-minimal head, %d encoded with a two-byte argument", arg)
		}
	case ai == 26:
		if r.remaining() < 4 {
			return 0, 0, 0, cborErr("truncated four-byte argument")
		}
		arg = uint64(r.b[r.i])<<24 | uint64(r.b[r.i+1])<<16 | uint64(r.b[r.i+2])<<8 | uint64(r.b[r.i+3])
		r.i += 4
		if arg < 65536 {
			return 0, 0, 0, cborErr("C4: non-minimal head, %d encoded with a four-byte argument", arg)
		}
	case ai == 27:
		if r.remaining() < 8 {
			return 0, 0, 0, cborErr("truncated eight-byte argument")
		}
		for k := 0; k < 8; k++ {
			arg = arg<<8 | uint64(r.b[r.i+k])
		}
		r.i += 8
		if arg < 1<<32 {
			return 0, 0, 0, cborErr("C4: non-minimal head, %d encoded with an eight-byte argument", arg)
		}
	case ai == 31:
		return 0, 0, 0, cborErr("C3: indefinite length is forbidden (major %d)", major)
	default: // 28, 29, 30
		return 0, 0, 0, cborErr("reserved additional information %d", ai)
	}
	return major, ai, arg, nil
}

func (r *cborReader) item(depth int) (Value, error) {
	if depth > MaxDepth {
		return Value{}, cborErr("C12: nesting depth %d exceeds MAX_DEPTH %d", depth, MaxDepth)
	}
	if r.remaining() < 1 {
		return Value{}, cborErr("truncated: expected an item")
	}

	// Простые значения разбираются до общей головы: профиль допускает ровно
	// два из них, и всё остальное в major 7 (float, null, undefined) отвергается.
	if r.b[r.i]>>5 == 7 {
		ib := r.b[r.i]
		r.i++
		switch ib & 0x1f {
		case 20:
			return Value{Kind: KindBool, Bool: false}, nil
		case 21:
			return Value{Kind: KindBool, Bool: true}, nil
		case 22:
			return Value{}, cborErr("C7: null is forbidden")
		case 23:
			return Value{}, cborErr("C7: undefined is forbidden")
		case 25, 26, 27:
			return Value{}, cborErr("C6: floats are forbidden")
		default:
			return Value{}, cborErr("C7: simple value 0x%02x is forbidden", ib)
		}
	}

	major, _, arg, err := r.head()
	if err != nil {
		return Value{}, err
	}

	switch major {
	case 0: // uint
		if arg > MaxUint {
			return Value{}, cborErr("uint %d exceeds MAX_UINT %d", arg, MaxUint)
		}
		return Value{Kind: KindUint, U: arg}, nil

	case 1: // negative
		return Value{}, cborErr("C8: negative integers are forbidden")

	case 2: // bstr
		if arg > MaxBstrBytes {
			return Value{}, cborErr("bstr of %d bytes exceeds MAX_BSTR_BYTES %d", arg, MaxBstrBytes)
		}
		n := int(arg)
		if n > r.remaining() {
			return Value{}, cborErr("bstr of %d bytes overruns the payload", n)
		}
		v := Value{Kind: KindBstr, B: r.b[r.i : r.i+n]}
		r.i += n
		return v, nil

	case 3: // tstr
		if arg > MaxTstrBytes {
			return Value{}, cborErr("tstr of %d bytes exceeds MAX_TSTR_BYTES %d", arg, MaxTstrBytes)
		}
		n := int(arg)
		if n > r.remaining() {
			return Value{}, cborErr("tstr of %d bytes overruns the payload", n)
		}
		raw := r.b[r.i : r.i+n]
		if !utf8.Valid(raw) {
			return Value{}, cborErr("C11: text string is not well-formed UTF-8")
		}
		r.i += n
		return Value{Kind: KindTstr, S: string(raw)}, nil

	case 4: // array
		if arg > MaxArrayItems {
			return Value{}, cborErr("array of %d items exceeds MAX_ARRAY_ITEMS %d", arg, MaxArrayItems)
		}
		n := int(arg)
		// Отказ до выделения: каждый элемент занимает минимум один байт.
		if n > r.remaining() {
			return Value{}, cborErr("array of %d items cannot fit in %d remaining bytes", n, r.remaining())
		}
		out := Value{Kind: KindArray, Array: make([]Value, 0, n)}
		for k := 0; k < n; k++ {
			it, err := r.item(depth + 1)
			if err != nil {
				return Value{}, err
			}
			out.Array = append(out.Array, it)
		}
		return out, nil

	case 5: // map
		if arg > MaxMapPairs {
			return Value{}, cborErr("map of %d pairs exceeds MAX_MAP_PAIRS %d", arg, MaxMapPairs)
		}
		n := int(arg)
		// Отказ до выделения: каждая пара занимает минимум два байта.
		if n > r.remaining()/2 {
			return Value{}, cborErr("map of %d pairs cannot fit in %d remaining bytes", n, r.remaining())
		}
		out := Value{Kind: KindMap, Map: make([]Pair, 0, n)}
		var prev uint64
		for k := 0; k < n; k++ {
			kmajor, _, karg, err := r.head()
			if err != nil {
				return Value{}, err
			}
			if kmajor != 0 {
				return Value{}, cborErr("C9: map key is major type %d, keys MUST be unsigned integers", kmajor)
			}
			if karg < CriticalKeyMin {
				return Value{}, cborErr("3.3: map key 0 is forbidden")
			}
			if karg > NonCriticalKeyMax {
				return Value{}, cborErr("3.3: map key %d is at or above 1024", karg)
			}
			if k > 0 && karg <= prev {
				return Value{}, cborErr("C10: map key %d does not exceed the previous key %d", karg, prev)
			}
			prev = karg
			val, err := r.item(depth + 1)
			if err != nil {
				return Value{}, err
			}
			out.Map = append(out.Map, Pair{Key: karg, Val: val})
		}
		return out, nil

	case 6: // tag
		return Value{}, cborErr("C5: tags are forbidden, tag %d seen", arg)
	}

	return Value{}, cborErr("unreachable major type %d", major)
}

File: apps/caramba-client/packages/caramba_vpn/lib/src/csm/codec/cbor.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:caramba_vpn/src/csm/errors.dart';

// Строгий профиль разбора CBOR, 03-WIRE.md 3. Каноничность здесь это
// локальный предикат над входящими байтами, а не перекодирование и сравнение.
// Декодер физически не умеет принять ничего, что профиль запрещает: нет
// неопределённых длин, нет тегов, нет чисел с плавающей точкой, нет
// отрицательных целых, нет null, нет строковых ключей карты.
//
// Ограничения проверяются по мере чтения заголовка, ДО выделения памяти,
// чтобы враждебный документ не мог заставить нас выделить буфер по его слову.

const int csmMaxDepth = 6;
const int csmMaxMapPairs = 64;
const int csmMaxArrayItems = 512;
const int csmMaxTstrBytes = 256;
const int csmMaxBstrBytes = 3072;
const int csmMaxUint = 9007199254740991; // 2^53 - 1

const int csmCriticalKeyMax = 63;
const int csmNonCriticalKeyMax = 1023;

/// Разобранное значение CBOR. Замкнутый набор: профиль допускает ровно эти
/// пять форм и никаких других.
sealed class CborValue {
  const CborValue();
}

class CborUint extends CborValue {
  const CborUint(this.value);
  final int value;
}

class CborBool extends CborValue {
  const CborBool(this.value);
  final bool value;
}

class CborBytes extends CborValue {
  const CborBytes(this.value);
  final Uint8List value;
}

class CborText extends CborValue {
  const CborText(this.value);
  final String value;
}

class CborArray extends CborValue {
  const CborArray(this.items);
  final List<CborValue> items;
}

class CborMap extends CborValue {
  const CborMap(this.entries);

  /// Пары в том порядке, в каком они пришли; порядок строго возрастающий по
  /// правилу C10, поэтому итерация ключей и есть отсортированный обход.
  final Map<int, CborValue> entries;

  CborValue? operator [](int key) => entries[key];
  bool has(int key) => entries.containsKey(key);
}

class _Reader {
  _Reader(this.bytes);

  final Uint8List bytes;
  int offset = 0;

  int get remaining => bytes.length - offset;

  Never fail(String detail) => csmFail(CsmErrorCode.parseCbor, 'P9', detail);

  int byte() {
    if (remaining < 1) {
      fail('payload ended inside a head');
    }
    return bytes[offset++];
  }

  /// Читает аргумент заголовка в кратчайшей форме (правило C4) и возвращает
  /// его значение. Неопределённая длина и значения 28..30 отвергаются здесь.
  int argument(int ai) {
    if (ai < 24) {
      return ai;
    }
    switch (ai) {
      case 24:
        final v = byte();
        if (v < 24) {
          fail('non-minimal one-byte argument for value $v');
        }
        return v;
      case 25:
        if (remaining < 2) {
          fail('payload ended inside a two-byte argument');
        }
        final v = (bytes[offset] << 8) | bytes[offset + 1];
        offset += 2;
        if (v < 256) {
          fail('non-minimal two-byte argument for value $v');
        }
        return v;
      case 26:
        if (remaining < 4) {
          fail('payload ended inside a four-byte argument');
        }
        final v = (bytes[offset] << 24) |
            (bytes[offset + 1] << 16) |
            (bytes[offset + 2] << 8) |
            bytes[offset + 3];
        offset += 4;
        if (v < 65536) {
          fail('non-minimal four-byte argument for value $v');
        }
        return v;
      case 27:
        if (remaining < 8) {
          fail('payload ended inside an eight-byte argument');
        }
        final hi = (bytes[offset] << 24) |
            (bytes[offset + 1] << 16) |
            (bytes[offset + 2] << 8) |
            bytes[offset + 3];
        final lo = (bytes[offset + 4] << 24) |
            (bytes[offset + 5] << 16) |
            (bytes[offset + 6] << 8) |
            bytes[offset + 7];
        offset += 8;
        if (hi == 0) {
          fail('non-minimal eight-byte argument');
        }
        // MAX_UINT это 2^53 - 1 именно для того, чтобы декодер на Dart или на
        // JavaScript никогда не терял точность молча.
        if (hi > 0x1fffff) {
          fail('unsigned integer above MAX_UINT');
        }
        return hi * 4294967296 + lo;
      case 31:
        fail('indefinite length is forbidden by rule C3');
      default:
        fail('reserved additional information $ai');
    }
  }

  Uint8List take(int n) {
    if (remaining < n) {
      fail('payload ended inside a string of $n bytes');
    }
    final out = Uint8List.sublistView(bytes, offset, offset + n);
    offset += n;
    return Uint8List.fromList(out);
  }
}

CborValue _value(_Reader r, int depth) {
  final ib = r.byte();
  final major = ib >> 5;
  final ai = ib & 0x1f;

  switch (major) {
    case 0:
      final v = r.argument(ai);
      if (v > csmMaxUint) {
        r.fail('unsigned integer above MAX_UINT');
      }
      return CborUint(v);
    case 1:
      r.fail('negative integers are forbidden by rule C8');
    case 2:
      final n = r.argument(ai);
      if (n > csmMaxBstrBytes) {
        r.fail('byte string of $n bytes exceeds MAX_BSTR_BYTES');
      }
      return CborBytes(r.take(n));
    case 3:
      final n = r.argument(ai);
      if (n > csmMaxTstrBytes) {
        r.fail('text string of $n bytes exceeds MAX_TSTR_BYTES');
      }
      final raw = r.take(n);
      try {
        return CborText(const Utf8Decoder().convert(raw));
      } on FormatException {
        r.fail('text string is not well-formed UTF-8');
      }
    case 4:
      if (depth > csmMaxDepth) {
        r.fail('nesting depth exceeds MAX_DEPTH');
      }
      final n = r.argument(ai);
      if (n > csmMaxArrayItems) {
        r.fail('array of $n items exceeds MAX_ARRAY_ITEMS');
      }
      final items = <CborValue>[];
      for (var i = 0; i < n; i++) {
        items.add(_value(r, depth + 1));
      }
      return CborArray(items);
    case 5:
      if (depth > csmMaxDepth) {
        r.fail('nesting depth exceeds MAX_DEPTH');
      }
      final n = r.argument(ai);
      if (n > csmMaxMapPairs) {
        r.fail('map of $n pairs exceeds MAX_MAP_PAIRS');
      }
      final entries = <int, CborValue>{};
      var previous = -1;
      for (var i = 0; i < n; i++) {
        final kb = r.byte();
        if ((kb >> 5) != 0) {
          r.fail('map key is not an unsigned integer, rule C9');
        }
        final key = r.argument(kb & 0x1f);
        // Правило C10 поглощает обнаружение дубликата в проверку порядка: одно
        // сравнение на ключ, и его нельзя забыть отдельно.
        if (key <= previous) {
          r.fail('map keys are not strictly ascending, rule C10');
        }
        previous = key;
        if (key == 0) {
          r.fail('map key 0 is rejected, 03-WIRE.md 3.3');
        }
        if (key > csmNonCriticalKeyMax) {
          r.fail('map key $key is at or above 1024, 03-WIRE.md 3.3');
        }
        entries[key] = _value(r, depth + 1);
      }
      return CborMap(entries);
    case 6:
      r.fail('tags are forbidden by rule C5');
    default:
      switch (ib) {
        case 0xf4:
          return const CborBool(false);
        case 0xf5:
          return const CborBool(true);
        case 0xf9:
        case 0xfa:
        case 0xfb:
          r.fail('floats are forbidden by rule C6');
        default:
          r.fail('simple value ${ib & 0x1f} is forbidden by rule C7');
      }
  }
}

/// Разбирает полезную нагрузку кадра как ровно один элемент CBOR верхнего
/// уровня, который обязан быть картой, и обязан израсходовать ровно
/// `payload_len` байт (правила C1 и C2).
CborMap csmDecodePayload(Uint8List payload) {
  final r = _Reader(payload);
  final v = _value(r, 1);
  if (r.remaining != 0) {
    csmFail(
      CsmErrorCode.parseCbor,
      'P9',
      'payload carries ${r.remaining} trailing bytes, rule C2',
    );
  }
  if (v is! CborMap) {
    csmFail(
      CsmErrorCode.parseCbor,
      'P9',
      'top-level item is not a map, rule C1',
    );
  }
  return v;
}
